File: api/app/services/FcmService.php
<?php

/**
 * class worker
 * отвечает за актуализацию задач для worker
 */

namespace services;

use model\SystemModel;
use model\AcceptorModel;
use model\UserModel;

class FcmService
{
    var $acceptor = null;
    var $push_system_id = null;
    var $user = null;


    function __construct($user_id)
    {
        $this->user = UserModel::get($user_id);
        $this->push_system_id = SystemModel::get_system_by_type('push')['id'];
        $this->acceptor = AcceptorModel::get_one_acceptor_by_system($this->push_system_id, $this->user);
    }

    private function get_tokens()
    {
        $fcm_tokens = [];

        if ($this->acceptor) {
            if ($this->acceptor['account']) {
                $fcm_tokens = explode(';', $this->acceptor['account']);
            }
        }

        return $fcm_tokens;
    }

    private function set_tokens($fcm_tokens)
    {
        $account_string = implode(";", $fcm_tokens);

        if ($this->acceptor) {
            $acceptor = $this->acceptor;
            $acceptor['account'] = $account_string;
            $acceptor = AcceptorModel::update_acceptor($acceptor['id'], $acceptor, $this->user);
        } else {
            $acceptor = array(
                'user_id' => $this->user['id'],
                'name' => 'Push',
                'system_id' => $this->push_system_id,
                'is_system' => true,
                'account' => $account_string
            );
            $acceptor = AcceptorModel::create_acceptor($acceptor, $this->user);
        }

        $this->acceptor = $acceptor;
    }

    public function add_fcm_token($new_fcm_token)
    {
        $fcm_tokens = $this->get_tokens();

        if (!in_array($new_fcm_token, $fcm_tokens)) {
            array_push($fcm_tokens, $new_fcm_token);
        }

        $this->set_tokens($fcm_tokens);
    }

    public function remove_fcm_token($token)
    {
        $fcm_tokens = $this->get_tokens();

        $to_remove = array($token);
        $new_fcm_tokens = array_diff($fcm_tokens, $to_remove);

        $this->set_tokens($new_fcm_tokens);
    }
}

File: backend/app/services/fcm_service.py
from typing import List

from sqlalchemy.orm import Session

from app.repo.crud.acceptor_crud import AcceptorCrud
from app.repo.crud.system_crud import SystemCrud
from app.repo.schemas.acceptor_scheme import AcceptorCreateScheme, AcceptorUpdateScheme
from app.services import ServiceResponse


class FcmService:
    def __init__(self, db: Session, user_id: int):
        self.acceptor_crud = AcceptorCrud(db=db)
        self.system_crud = SystemCrud(db=db)

        self.user_id = user_id
        self.push_system_id = self.system_crud.get_by_type(type='push').id
        self.acceptor = self.acceptor_crud.get_by_user_id_and_system_id(user_id=user_id, system_id=self.push_system_id)

    def _get_tokens(self):
        fcm_tokens = []

        if self.acceptor:
            if self.acceptor.account:
                fcm_tokens = self.acceptor.account.split(';')

        return fcm_tokens

    def set_tokens(self, fcm_tokens: List) -> ServiceResponse:
        account_string = ';'.join(fcm_tokens)

        if self.acceptor:
            acceptor_scheme = AcceptorUpdateScheme(account=account_string)
            self.acceptor_crud.update(db_object=self.acceptor, scheme=acceptor_scheme)
        else:
            acceptor_scheme = AcceptorCreateScheme(
                user_id=self.user_id,
                name='Push',
                system_id=self.push_system_id,
                is_system=True,
                account=account_string,
                is_disabled=False
            )
            self.acceptor = self.acceptor_crud.create(scheme=acceptor_scheme)

        return ServiceResponse()

    def add_fcm_token(self, token: str) -> ServiceResponse:
        fcm_tokens = self._get_tokens()

        if token not in fcm_tokens:
            fcm_tokens.append(token)

        self.set_tokens(fcm_tokens)

        return ServiceResponse()

    def remove_fcm_token(self, token: str) -> ServiceResponse:
        fcm_tokens = self._get_tokens()
        to_remove = [token]
        new_fcm_tokens = list(set(fcm_tokens) - set(to_remove))

        self.set_tokens(new_fcm_tokens)

        return ServiceResponse()
